src/composable/orderTypes/Twap.ts

```diff
@@ -273,13 +273,21 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
     }
 
     const cabinet = await this.cabinet(params)
-    const cabinetEpoch = utils.defaultAbiCoder.decode(['uint256'], cabinet)[0]
+    const rawCabinetEpoch = utils.defaultAbiCoder.decode(['uint256'], cabinet)[0] as BigNumber
+
+    // Guard against out-of-range cabinet epoch
+    if (rawCabinetEpoch.gt(MAX_UINT32)) {
+      throw new Error(`Cabinet epoch out of range: ${rawCabinetEpoch.toString()}`)
+    }
+
+    // Convert the cabinet epoch (bignumber) to a number.
+    const cabinetEpoch = rawCabinetEpoch.toNumber()
 
     if (cabinetEpoch === 0) {
       throw new Error('Cabinet is not set. Required for TWAP orders that start at mining time.')
     }
 
-    return parseInt(cabinet, 16)
+    return cabinetEpoch
```

Comment:
What about, instead of converting it to a `number` and potentially lose precision, work instead with a BigNumber instance?

Unless, we know this should never exceed default JS number precision, and thus the MAX_UINT32 check above.

Reply:
In general, I'd agree with the `BigNumber` approach. The main reason I kept with the `number` is if it's a legitimate TWAP, the TWAP expects a `block.timestamp` to be in the cabinet. As any arbitrary value can _technically_ be in the cabinet (they mess with the `createWithContext` call), it is not _guaranteed_, in which case it will throw and be classified as a malformed / invalid order, otherwise, we know for sure that it will never exceed the JS number precision.